Lib/test/test_dataclasses.py

```diff
@@ -2902,6 +2902,49 @@ class A:
         self.assertEqual(obj.a, 'a')
         self.assertEqual(obj.b, 'b')
 
+    def test_frozen_slots_raises_correct_error(self):
+        # bpo-45897
+        @dataclass(slots=True, frozen=True)
+        class Point:
+            x: int
+            y: int
+
+        p = Point(1, 2)
+        self.assertEqual(p.x, 1)
+        self.assertEqual(p.y, 2)
+        with self.assertRaises(FrozenInstanceError):
+            p.x = 2
+
+        try:
+            p.z = 5
+        except Exception as exc:
+            exc_type = type(exc)
+            if exc_type is not AttributeError:
+                raise TypeError(
+                    f'AttributeError was expected; got {exc_type.__name__!r} instead'
+                ) from exc
+            else:
+                # Good! The test passes; AttributeError is what we want
+                pass
+        else:
+            self.fail('AttributeError expected, but no error was raised')
```

Comment:
But why not assertRaises() like for `p.x = 2` above?

```python
        with self.assertRaises(AttributeError):
            p.z = 5
```

Or, at least, not non-nested exceptions?

```python
        try:
            p.z = 5
        except AttributeError:
            # Good! The test passes; AttributeError is what we want
            pass
        except  Exception as exc:
            raise TypeError(
                f'AttributeError was expected; got {type(exc).__name__!r} instead'
            ) from exc
```

Reply:
`FrozenInstanceError` is a subclass of `AttributeError`, so `self.assertRaises(AttributeError)` or `except AttributeError` will pass if a `FrozenInstanceError` is raised. But we don't *want* a `FrozenInstanceError` in this case — we want a plain-old `AttributeError`.

Reply:
In other words: we want to verify that this error is being raised by the standard machinery that prevents new attributes being added that aren't specified in `__slots__`. We want to make sure that this error is _not_ being caused by any special `dataclass` shenanigans.

Reply:
I think it would be clearer to use assertRaises, then check the type of the exception.
```
class E(AttributeError): pass

class TestCase(unittest.TestCase):
    def test(self):
        with self.assertRaises(AttributeError) as ex:
            raise E()
        # Look for an exact AttributeError, not a subclass of it.
        self.assertIs(type(ex.exception), AttributeError,
                      f"expected 'AttributeError', got {type(ex.exception).__name__!r}")
```
I was also confused by the existing code. In any event, I think a comment is definitely warranted.

Reply:
I'll make the change.

Reply:
How does it look now?

Reply:
It looks good, thanks. I haven't had time to do a proper review of the rest of the code.